File: Offer 24. 反转链表.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 迭代
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null, curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}

// 递归
class Solution {
    public ListNode reverseList(ListNode head) {
        // 至少需要两个节点
        if (head == null || head.next == null) {
            return head;
        }

        // 递归部分
        ListNode newHead = reverseList(head.next);
        // 修改当前位置的指向
        head.next.next = head;
        head.next = null;

        return newHead;
    }
}

File: Offer 24. 反转链表.py

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 递归
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head        
        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head


# 迭代
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        cur, pre = head, None
        while cur:
            tmp = cur.next
            cur.next = pre
            pre = cur
            cur = tmp
        return pre